### PR TITLE
sql: add a starts_with function

### LINKED ISSUE #29520: surfaces/adapter: support `starts_with` PostgreSQL function
### Feature request

Consider supporting the [`starts_with`](https://www.postgresql.org/docs/16/functions-string.html) function, which determines whether the provided string starts with the specified prefix. This function was introduced in PostgreSQL 11 ([commit](https://git.postgresql.org/gitweb/?p=postgresql.git;a=commit;h=710d90da1fd8c1d028215ecaf7402062079e99e9)).

```sql
postgres=# SELECT starts_with('foobar', 'foo');
 starts_with 
-------------
 t
(1 row)
```

cc @matthelm

### PR DESCRIPTION
**Pull request by benesch**
Fix #29520.

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Review by petrosagg — APPROVED** [review]
LGTM! I was expecting to see the entry in `propagates_nulls` and I was surprised to see that we don't use an exhaustive match there! Looks like a potential footgun

File: doc/user/data/sql_funcs.yml [hunk 1]
```diff
@@ -416,6 +416,9 @@
   - signature: 'split_part(s: str, d: s, i: int) -> str'
     description: Split `s` on delimiter `d`. Return the `str` at index `i`, counting from 1.
 
+  - signature: 'starts_with(s: str, prefix: str) -> str'
```
Comment (matthelm) [inline doc/user/data/sql_funcs.yml:419]: Is this the correct return type?

Comment (benesch) [reply doc/user/data/sql_funcs.yml]: Ah, thanks. Fixed.

**Review by ParkMyCar — APPROVED** [review]
sweet!

File: src/sql/src/func.rs [hunk 2]
```diff
@@ -3847,6 +3847,9 @@ pub static MZ_CATALOG_BUILTINS: LazyLock<BTreeMap<&'static str, Func>> = LazyLoc
             params!(String) => UnaryFunc::SeahashString(func::SeahashString) => UInt32, oid::FUNC_SEAHASH_STRING_OID;
             params!(Bytes) => UnaryFunc::SeahashBytes(func::SeahashBytes) => UInt32, oid::FUNC_SEAHASH_BYTES_OID;
         },
+        "starts_with" => Scalar {
+            params!(String, String) => BinaryFunc::StartsWith => Bool, 3696;
```
Comment (antiguru) [inline src/sql/src/func.rs:3851]: Might make sense to define an `oid` constant here!

**Review by antiguru — APPROVED** [review]
LGTM, left a comment.

File: src/sql/src/func.rs (hunk 2 above)
Comment (benesch) [reply src/sql/src/func.rs:3851]: AFAICT we only introduce OID constants for Materialize-specific functions/types; OIDs that come from PostgreSQL are just inlined in this file. 🤷🏽

**Pull request merged**